### PR TITLE
Add missing implicit SFINAE constraints

### LINKED ISSUE #214: STL: "Effects: Equivalent to" implies Constraints
The Standard frequently uses "*Effects:* Equivalent to" when specifying things. This is awesome because it provides an implementation that can be used in production code, nearly unchanged, 99% of the time. (Occasionally, a more efficient implementation is possible.)

For example, `shared_ptr`'s assignment from `unique_ptr` [is specified as](https://eel.is/c++draft/util.smartptr.shared.assign#6):

> *Effects:* Equivalent to `shared_­ptr(std​::​move(r)).swap(*this)`.

However, `shared_ptr`'s constructor from `unique_ptr` [is constrained](https://eel.is/c++draft/util.smartptr.shared.const#28) (currently with old-style wording): 

> *Remarks:* This constructor shall not participate in overload resolution unless `Y*` is compatible with `T*` and `unique_­ptr<Y, D>​::​pointer` is convertible to `element_­type*`.

Finally, the Standard's Library-wide wording [says](https://eel.is/c++draft/description#structure.specifications-4):

> Whenever the *Effects* element specifies that the semantics of some function `F` are *Equivalent to* some code sequence, then the various elements are interpreted as follows. If `F`'s semantics specifies any *Constraints* or *Mandates* elements, then those requirements are logically imposed prior to the *equivalent-to* semantics.

We've never implemented that. For example, we constrain the constructor:

https://github.com/microsoft/STL/blob/6b0238d7035236af0f81917262691db701e70fe3/stl/inc/memory#L1109-L1113

But we **don't** constrain the assignment operator:

https://github.com/microsoft/STL/blob/6b0238d7035236af0f81917262691db701e70fe3/stl/inc/memory#L1162-L1166

This is a bug because the difference is detectable through `is_assignable`.

The extent of this problem is unknown. I don't know of a better way to thoroughly fix this issue than going through every occurrence of "*Effects:* Equivalent to" in the Standard and looking for expressions that involve constrained machinery.

### LINKED ISSUE #2299: `<memory>`, `<optional>`, `<any>`: Implicit SFINAE constraints
**Describe the bug**
According to [[structure.specifications]/4](https://eel.is/c++draft/structure.specifications#4), some uses of "_Effects:_ Equivalent to" specifies "_Constraints:_" of the underlying function calls. I suppose that it only propagates "_Constraints:_" specified for standard library functions, not literal `requires` or other substitution failures, and the implied "_Constraints:_" were specified by "_Remarks:_" in C++11~17 (in other words, essentially unchanged in C++20).

From a quick glance at WG21-N4901, I think the following functions have additional SFINAE constraints (and currently there're some defects in N4901):
- template overloads of `std::shared_ptr::operator=`
- template overloads of `std::shared_ptr::reset`
- template overloads of `std::weak_ptr::operator=`
- `std::make_optional`
- `std::make_any`
- the template overload of `std::move_only_function::operator=`
(Anything missing?)

Currently none of these functions is constrained in MSVC STL (except for `std::move_only_function::operator=`, which is being implemented).

As all of these functions have the same "_Constraints:_" as the selected constructors, perhaps we can simply add `enable_if_t<is_constructible_v<...>, int> = 0` to them.

I have submitted 2 related LWG issues:
- LWG-3627: `std::make_optional` overloads are inconsitently specified. IMO they should all be constrained.
- LWG-3628: `uninitialized` algorithms and some `std::reduce` overloads are also specified by "_Effects:_ Equivalent to", but I feel that they are misspecified and should not have additional SFINAE constraints.

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #214 
Fixes #2299 
Also speculatively implements LWG-3627.

### CODE REVIEW

File: stl/inc/any [hunk 1]
```diff
@@ -357,11 +357,13 @@ inline void swap(any& _Left, any& _Right) noexcept {
     _Left.swap(_Right);
 }
 
-template <class _ValueType, class... _Types>
+template <class _ValueType, class... _Types,
+    enable_if_t<is_constructible_v<any, in_place_type_t<_ValueType>, _Types...>, int> = 0>
```
Comment (StephanTLavavej) [inline stl/inc/any:361]: No change requested: I observe that `in_place_type_t<_ValueType>` doesn't match the constness and value category of the function argument, but it doesn't matter (we control the overload set enough). @CaseyCarter agrees that the shorter form is better here.

File: tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+using namespace std;
+
+// tests for shared_ptr<T>::operator=
+template <class T, class U, class = void>
+constexpr bool can_shared_ptr_assign = false;
+
+template <class T, class U>
+constexpr bool can_shared_ptr_assign<T, U, void_t<decltype(declval<shared_ptr<T>&>() = declval<U>())>> = true;
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp:25]: Technically should include `<utility>` for `declval`.

File: tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+template <class T, class... Us>
+constexpr bool can_make_optional_impl<void_t<decltype(make_optional<T>(declval<Us>()...))>, T, Us...> = true;
+
+template <class T, class... Us>
+constexpr bool can_make_optional_usual = can_make_optional_impl<void, T, Us...>;
+
+STATIC_ASSERT(!can_make_optional_decay<const unique_ptr<int>&>); // LWG-3627
+STATIC_ASSERT(!can_make_optional_usual<int, int, int>);
+STATIC_ASSERT(!can_make_optional_usual<int, initializer_list<int>&>);
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp:79]: Technically should include `<initializer_list>`.

File: tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+#include <type_traits>
+#include <version>
+
+#ifdef __cpp_lib_optional
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp:8]: While this is conforming, our tests conventionally use the coarse-grained (and implementation-specific, of course) `_HAS_CXX17` etc. macros, when the presence of a feature doesn't vary in other ways (e.g. varying with compilers). The product code also follows this convention (e.g. we had a lot of fine-grained tests for `__cpp_lib_constexpr_dynamic_alloc` that were eventually unified into `_HAS_CXX20`). It's not a big deal, but preferring the coarse-grained macros makes it a bit easier to see what will be tested as the Standard mode varies, and it makes the remaining uses of the fine-grained feature-test macros stand out as being special.

File: tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+template <class T, class... Us>
+constexpr bool can_make_optional_impl<void_t<decltype(make_optional<T>(declval<Us>()...))>, T, Us...> = true;
+
+template <class T, class... Us>
+constexpr bool can_make_optional_usual = can_make_optional_impl<void, T, Us...>;
+
+STATIC_ASSERT(!can_make_optional_decay<const unique_ptr<int>&>); // LWG-3627
+STATIC_ASSERT(!can_make_optional_usual<int, int, int>);
+STATIC_ASSERT(!can_make_optional_usual<int, initializer_list<int>&>);
+#endif
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp:80]: We conventionally comment `#endif` directives (unless the controlled region is extremely short and distant from other directives), which makes it easier to see how they match up.

File: tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+template <class T, class U>
+constexpr bool can_shared_ptr_assign<T, U, void_t<decltype(declval<shared_ptr<T>&>() = declval<U>())>> = true;
+
+STATIC_ASSERT(!can_shared_ptr_assign<int, shared_ptr<long>>);
+STATIC_ASSERT(!can_shared_ptr_assign<int, const shared_ptr<long>&>);
+STATIC_ASSERT(!can_shared_ptr_assign<int, unique_ptr<long>>);
+#if _HAS_AUTO_PTR_ETC
+STATIC_ASSERT(!can_shared_ptr_assign<int, auto_ptr<long>>);
+#endif
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002299_implicit_sfinae_constraints/test.compile.pass.cpp:32]: This is good, but I recommend having one `STATIC_ASSERT` where the handwritten type trait is `true` - this verifies that the type trait works (and isn't always `false` for some other reason).

Occurs below.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! Product code looked perfect, so I went ahead and pushed minor test changes (the most significant of which was ensuring that the custom type traits can return both `true` and `false`).

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving the library's conformance to this invisible Standardese! :joy_cat: :eyeglasses: :heart_eyes_cat:

**Pull request merged**